docs/CONTRIBUTING.md

```diff
@@ -0,0 +1,246 @@
+
+- Review and follow the Style Guide (_Coming Soon_)
+- Write tests for your code. All new features and functionality must have tests demonstrating correctness
+  - Update existing tests if you change functionality
+  - _New contributors: A great way to better understand the codebase is to try writing tests for existing code_
+- Keep commits focused and meaningful
+  - Each commit should represent one logical change
+  - Write clear commit messages that describe that change ([See guidelines](#writing-good-commit-messages))
+- Update documentation as you go
+  - Add JSDoc/JavaDoc comments to new code
+    - _New contributors: Another great way to learn the codebase is to document existing code_
+  - Update repo Markdown docs if needed
+  - Code should for the most part be self-documenting. If not, make sure to add inline comments
+- Push your code to GitHub at the end of each coding session. This lets other
+  team members see how you're doing
+
+#### Writing Good Commit Messages
+- Start with a present-tense verb that describes what the commit does:
+    - `add late submission validation`
+    - `fix grade calculation bug`
+    - `update setup instructions`
+    - `remove unused imports`
+    - `refactor test runner`
+    - `document API endpoints`
+- Use lowercase letters
+- Keep it concise but clear (aim for under 50 characters)
+- Bad examples:
+    - `changes` (too vague)
+    - `Added new feature` (past tense)
+    - `WIP` (uninformative)
+    - `Fix stuff` (too vague)
+
+For bigger changes, you can add more details after the first line.
+
+> [!TIP]
+> Remember: A good commit message lets other developers know what changed
+without having to look at the code!
+
+### 4. Pull Requests
+Pull requests are how you get your changes merged into the codebase. 
+You create a pull request once you feel your changes are ready.
+
+#### Creating a Pull Request
+- Make sure you have pushed all changes to GitHub first
+- The title of your pull request should be simple, concise, and descriptive
+  - It should have a prefix just like an issue
+  - It can be basically the name of the branch
+    - Issue: `Frontend: refactor all server calls to one place`
+    - Branch: `server-communicator`
+    - Pull Request: `Frontend: ServerCommunicator`
+- Fill out the Pull Request template with enough detail for others to understand what 
+  changed without having to read all the new code.
+- Make sure all GitHub actions pass:
+  - Your pull request will be blocked from merging if all tests don't pass
+- Move the connected issue into the `PR Submittted` column on the 
+  [TA Project Board](https://github.com/orgs/softwareconstruction240/projects/1/views/9)
+- Request a review from other Autograder developers
+  - If you're changing an existing system, request a review from the developer that 
+    wrote it initially, or is most familiar with that system
+  - You can request multiple reviewers, but you're only required to get approval from one
```

Comment:
Maybe this is me being tyrannical (and please tell me if you disagree), but I would like to wait for a review from the admin before merging (specifically if the pull request is to main--if its to a different branch then just the owner of the other branch would be necessary). Technically GitHub doesn't require that, but it might be good.

Reply:
Actually I think GitHub can enforce that, if we make you/the lead TA/whoever the `CODEOWNER` of everything in the repo. According to GPT there's a **Require review from Code Owners** option. And no, I don't think that's tyrannical. If anything, my concern is that it might become a bottleneck, but as long as the admin is on top of things, I think it's a good idea.

Reply:
I love and support this discussion!

Reply:
NOTE: as of the dev meeting, this is outdated.